spec/unit/git/rugged/credentials_spec.rb

```diff
@@ -0,0 +1,76 @@
+
+  let(:repo) { R10K::Git::Rugged::BareRepository.new("/some/nonexistent/path", "repo.git") }
+
+  subject { described_class.new(repo) }
+
+  describe "determining the username" do
+    before { R10K::Git.settings[:username] = "moderns" }
+    after { R10K::Git.settings.reset! }
+
+    it "prefers a username from the URL" do
+      user = subject.get_git_username("https://tessier-ashpool.freeside/repo.git", "ashpool")
+      expect(user).to eq "ashpool"
+    end
+
+    it "uses the username from the Git config when specified" do
+      user = subject.get_git_username("https://tessier-ashpool.freeside/repo.git", nil)
+      expect(user).to eq "moderns"
+    end
+
+    it "falls back to the current user" do
+      R10K::Git.settings.reset!
+      expect(Etc).to receive(:getlogin).and_return("finn")
+      user = subject.get_git_username("https://tessier-ashpool.freeside/repo.git", nil)
+      expect(user).to eq "finn"
+    end
+  end
+
+  describe "generating ssh key credentials" do
+    after { R10K::Git.settings.reset! }
+
+    it "raises an error if no key has been set" do
+      R10K::Git.settings[:private_key] = nil
+      expect {
+        subject.get_ssh_key_credentials("https://tessier-ashpool.freeside/repo.git", nil)
+      }.to raise_error(R10K::Git::GitError, /no private key was given/)
+    end
+
+    it "generates the rugged sshkey credential type" do
+      R10K::Git.settings[:private_key] = "/some/nonexistent/.ssh/key"
+      creds = subject.get_ssh_key_credentials("https://tessier-ashpool.freeside/repo.git", nil)
+      expect(creds).to be_a_kind_of(Rugged::Credentials::SshKey)
+      expect(creds.instance_variable_get(:@privatekey)).to eq("/some/nonexistent/.ssh/key")
+    end
+  end
+
+  describe "generating plaintext credentials" do
+    it "generates the rugged user/password credential type" do
+      creds = subject.get_plaintext_credentials("https://azurediamond:hunter2@tessier-ashpool.freeside/repo.git", "azurediamond")
+      expect(creds).to be_a_kind_of(Rugged::Credentials::UserPassword)
+      expect(creds.instance_variable_get(:@username)).to eq("azurediamond")
+      expect(creds.instance_variable_get(:@password)).to eq("hunter2")
+    end
+  end
+
+  describe "generating credentials" do
+    it "creates ssh key credentials for the sshkey allowed type" do
+      expect(subject).to receive(:get_ssh_key_credentials).with("https://tessier-ashpool.freeside/repo.git", nil).and_return :sshkey
+      expect(subject.call("https://tessier-ashpool.freeside/repo.git", nil, [:ssh_key])).to eq :sshkey
+    end
+
+    it "creates user/password credentials for the default allowed type" do
+      expect(subject).to receive(:get_plaintext_credentials).with("https://tessier-ashpool.freeside/repo.git", nil).and_return :userpassword
+      expect(subject.call("https://tessier-ashpool.freeside/repo.git", nil, [:plaintext])).to eq :userpassword
+    end
+
+    it "creates default credentials when no other types are allowed" do
```

Comment:
perhaps these tests are mocking too much out since they hid the lack of this method.